### PR TITLE
Add popup to extend expiring sessions

### PR DESCRIPTION
**Pull request by munnsmunns**
When a user's session cookie is within 15 minutes of expiring, they will receive a popup requesting them to click OK in order to extend the session. If they click the popup, we'll extend the current user session for an additional duration of a session. If they do not click, and then click after the session has already expired, they'll return to the homepage.

### CODE REVIEW

File: frontend/shared/startup.js [hunk 1]
```diff
@@ -29,6 +29,26 @@ const getCookie = function(name) {
     sessionid = getCookie("sessionid"),
     csrfSafeMethod = method => /^(GET|HEAD|OPTIONS|TRACE)$/.test(method);
 
+var intervalID;
```
Comment (shapiromatron) [inline frontend/shared/startup.js:32]: could you move this code into a separate file,  for example, see the as an example. That way all the variable closures will a single place

```javascript
import sidebarStartup from "./utils/sidebarStartup";
```

File: frontend/shared/utils/checkSession.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+            if (expTime - Date.now() < SESSION_EXPIRE_WARNING) {
+                const resp = confirm(
+                    "Your current session will expire in 15 minutes. Please click OK to stay logged in."
+                );
+                if (resp) {
+                    // confirmed request; cancel this interval check and make request
+                    clearInterval(checkId); // prevent popup stacking on top of each other
+                    $.post("/update-session/", {refresh: true}).done(response => {
+                        // session update failed; go home
+                        if (response.new_expiry_time === null) {
```
Comment (shapiromatron) [inline frontend/shared/utils/checkSession.js:26]: i added this check in for cases where a user may press "ok" after the session has already expired

File: frontend/shared/utils/checkSession.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
+const checkSession = function() {
```
Comment (shapiromatron) [inline frontend/shared/utils/checkSession.js:1]: did a little reorganization of this function, hopefully it reads ok

File: frontend/shared/utils/checkSession.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+                        }
+                        setExpireTime(response.new_expiry_time);
+                        checkId = setInterval(check, CHECK_INTERVAL_MS);
+                    });
+                }
+            }
+        };
+
+    // only setup check if we have a valid session expiration
+    if (getExpireTime()) {
```
Comment (shapiromatron) [inline frontend/shared/utils/checkSession.js:38]: added a check here; we only want to set this interval if the page reports a valid session duration

File: hawc/apps/assessment/views.py [hunk 5]
```diff
@@ -666,9 +666,26 @@ def isTruthy(self, request, field):
...
+            if request.user.is_authenticated:
+                old_time = request.session.get_expiry_date().isoformat()
+                request.session.set_expiry(None)  # use the global session expiry policy
+                new_time = request.session.get_expiry_date().isoformat()
+                response = {
+                    "message": f"Session extended from {old_time} to {new_time}.",
+                    "new_expiry_time": new_time,
+                }
+            else:
+                response = {
```
Comment (shapiromatron) [inline hawc/apps/assessment/views.py:684]: we hit this case if a request isn't bound to a user, which happens when a session expires

**Review by shapiromatron — APPROVED** [review]
I made some changes to the code but nothing major; I was able to color within the edges you set up. Nice job. Please give it a test to make sure it still performs as you'd expect, but if it works, ok to merge!

**Comment by munnsmunns** [comment]
> I made some changes to the code but nothing major; I was able to color within the edges you set up. Nice job. Please give it a test to make sure it still performs as you'd expect, but if it works, ok to merge!

Looks good! You covered a couple scenarios I didn't think about.

**Pull request merged**